src/state.rs

```diff
@@ -308,88 +155,42 @@ pub struct Allowance {
 impl Allowance {
     pub fn is_expired_at(&self, block: &cosmwasm_std::BlockInfo) -> bool {
         match self.expiration {
-            Some(time) => block.time >= time,
+            Some(time) => block.time.seconds() >= time,
             None => false, // allowance has no expiration
         }
     }
 }
 
-pub fn read_allowance<S: Storage>(
-    store: &S,
-    owner: &CanonicalAddr,
-    spender: &CanonicalAddr,
-) -> StdResult<Allowance> {
-    let owner_store =
-        ReadonlyPrefixedStorage::multilevel(&[PREFIX_ALLOWANCES, owner.as_slice()], store);
-    let owner_store = TypedStore::attach(&owner_store);
-    let allowance = owner_store.may_load(spender.as_slice());
-    allowance.map(Option::unwrap_or_default)
-}
-
-pub fn write_allowance<S: Storage>(
-    store: &mut S,
-    owner: &CanonicalAddr,
-    spender: &CanonicalAddr,
-    allowance: Allowance,
-) -> StdResult<()> {
-    let mut owner_store =
-        PrefixedStorage::multilevel(&[PREFIX_ALLOWANCES, owner.as_slice()], store);
-    let mut owner_store = TypedStoreMut::attach(&mut owner_store);
-
-    owner_store.store(spender.as_slice(), &allowance)
-}
-
-// Viewing Keys
-
-pub fn write_viewing_key<S: Storage>(store: &mut S, owner: &CanonicalAddr, key: &ViewingKey) {
-    let mut balance_store = PrefixedStorage::new(PREFIX_VIEW_KEY, store);
-    balance_store.set(owner.as_slice(), &key.to_hashed());
-}
+pub static ALLOWANCES: Keymap<(Addr, Addr), Allowance> = Keymap::new(PREFIX_ALLOWANCES);
+pub struct AllowancesStore {}
+impl AllowancesStore {
+    pub fn load(store: &dyn Storage, owner: &Addr, spender: &Addr) -> Allowance {
+        ALLOWANCES
+            .get(store, &(owner.clone(), spender.clone()))
+            .unwrap_or_default()
+    }
 
-pub fn read_viewing_key<S: Storage>(store: &S, owner: &CanonicalAddr) -> Option<Vec<u8>> {
-    let balance_store = ReadonlyPrefixedStorage::new(PREFIX_VIEW_KEY, store);
-    balance_store.get(owner.as_slice())
+    pub fn save(
+        store: &mut dyn Storage,
+        owner: &Addr,
+        spender: &Addr,
+        allowance: &Allowance,
+    ) -> StdResult<()> {
+        ALLOWANCES.insert(store, &(owner.clone(), spender.clone()), allowance)
+    }
 }
 
 // Receiver Interface
 
-pub fn get_receiver_hash<S: ReadonlyStorage>(
-    store: &S,
-    account: &HumanAddr,
-) -> Option<StdResult<String>> {
-    let store = ReadonlyPrefixedStorage::new(PREFIX_RECEIVERS, store);
+pub fn get_receiver_hash(store: &dyn Storage, account: &Addr) -> Option<StdResult<String>> {
+    let store = ReadonlyPrefixedStorage::new(store, PREFIX_RECEIVERS);
```

Comment:
Can we simplify by usin `Item` here?

Reply:
I don't think so because for each account we need its receiver hash.
Item will only save one item per prefix

Reply:
Why not:
```rust
pub const RECEIVER_HASH = Item<String>::new(PREFIX_RECEIVERS);

let receiver_hash = RECEIVER_HASH.add_suffix(account.as_str().as_bytes())?;
```

Reply:
ok